### PR TITLE
http: adding an option to preserve hop by hop headers

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low (config guarded code)
Testing: new unit tests
Docs Changes: n/a
Release Notes: intentionally not added
Platform Specific Features: n/a
Part of https://github.com/envoyproxy/envoy-mobile/issues/1540

### CODE REVIEW

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 1]
```diff
@@ -371,6 +371,11 @@ message HttpConnectionManager {
   ServerHeaderTransformation server_header_transformation = 34
       [(validate.rules).enum = {defined_only: true}];
 
+  // By default, Envoy removes hop by hop headers. For envoy-mobile
+  // they should be passed up to the caller. This should not be used by server
+  // side Envoy as it expects hop by hop headers to be cleared.
+  bool preserve_hop_by_hop_response_headers = 48;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:377]: Thinking out loud here, are there going to be more mobile-only fields? If so, should this be part of a sub message that makes it clear that this is only for envoy-mobile instances?
This way we can also reject a config that has this sub-message in non-envoy-mobile instances.

**Review by adisuissa — COMMENTED** [review]
Got a question about this.
/wait-any

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:377]: Also for consideration, can we make HCM aware of, in some general sense, the fact that it's linked into a client and not another hop? I think to reject and validate this field on Envoy-proper, we would need this kind of awareness anyway.

**Comment by alyssawilk** [comment]
@mattklein123 @goaway will need unit tests but WDYT of this vs explicit control knob?

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 2]
```diff
@@ -222,6 +222,35 @@ HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped(
...
+  // Keep in mind the lambda capture list **doesn't** determine the destruction order, but it's fine
+  // as these captured objects are also global singletons.
+  return [singletons, filter_config, &context](Network::FilterManager& filter_manager) -> void {
+    auto hcm = new Http::ConnectionManagerImpl(
+        *filter_config, context.drainDecision(), context.api().randomGenerator(),
+        context.httpContext(), context.runtime(), context.localInfo(), context.clusterManager(),
+        context.overloadManager(), context.dispatcher().timeSource());
+    hcm->setClearHopByHopResponseHeaders(false);
+    filter_manager.addReadFilter(Network::ReadFilterSharedPtr{hcm});
+  };
```
Comment (mattklein123) [inline source/extensions/filters/network/http_connection_manager/config.cc:251]: nit: I think you can probably share this code in a utility function with above.

**Review by mattklein123 — COMMENTED** [review]
Yup this is exactly what I had in mind. Thank you!

**Comment by alyssawilk** [comment]
I think macos failure is spurious, PTAL!

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 3]
```diff
@@ -1000,3 +1000,9 @@ message RequestIDExtension {
   // Request ID extension specific configuration.
   google.protobuf.Any typed_config = 1;
 }
+
+// [#protodoc-title: Envoy Mobile HTTP connection manager]
+// HTTP connection manager for use in Envoy mobile.
+message EnvoyMobileHttpConnectionManager {
+  HttpConnectionManager config = 1;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:1007]: I guess it's pretty trivial but a comment should be added for this field.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by alyssawilk** [comment]
@snowp any requests here?  I think I addressed all @mattklein123's comments

File: source/common/http/conn_manager_utility.cc [hunk 4]
```diff
@@ -436,19 +437,21 @@ void ConnectionManagerUtility::mutateResponseHeaders(ResponseHeaderMap& response
...
     config.requestIDExtension()->setInResponse(response_headers, *request_headers);
   }
-  response_headers.removeKeepAlive();
-  response_headers.removeProxyConnection();
-
+  if (clear_hop_by_hop) {
+    response_headers.removeTransferEncoding();
+    response_headers.removeKeepAlive();
+    response_headers.removeProxyConnection();
+  }
```
Comment (snowp) [inline source/common/http/conn_manager_utility.cc:454]: Any reason to not do this within one if block?

File: source/common/http/conn_manager_utility.h [hunk 5]
```diff
@@ -74,10 +71,25 @@ class ConnectionManagerUtility {
...
-
+  /**
+   * Mutates response headers in various ways. This functionality is broken out because of its
+   * complexity for ease of testing. See the method itself for detailed comments on what
+   * mutations are performed.
+   *
+   * Note this function may be called twice on the response path if there are
+   * 100-Continue headers.
+   *
+   * @param response_headers the headers to sanitize.
```
Comment (snowp) [inline source/common/http/conn_manager_utility.h:82]: nit: `headers to mutate.`? To match the function name

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 6]
```diff
@@ -214,14 +222,28 @@ HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped(
   // reference count.
   // Keep in mind the lambda capture list **doesn't** determine the destruction order, but it's fine
   // as these captured objects are also global singletons.
-  return [singletons, filter_config, &context](Network::FilterManager& filter_manager) -> void {
-    filter_manager.addReadFilter(Network::ReadFilterSharedPtr{new Http::ConnectionManagerImpl(
+  return [singletons, filter_config, &context,
+          clear_hop_by_hop_headers](Network::FilterManager& filter_manager) -> void {
+    auto hcm = new Http::ConnectionManagerImpl(
```
Comment (snowp) [inline source/extensions/filters/network/http_connection_manager/config.cc:227]: make_shared?

**Review by snowp — COMMENTED** [review]
Just a few small comments, otherwise this LG!

File: source/extensions/filters/network/http_connection_manager/config.cc (hunk 6 above)
Comment (alyssawilk) [reply source/extensions/filters/network/http_connection_manager/config.cc:227]: We could make the shared readfiltersharedptr but then would need to cast it back for the call to setClearHopByHopResponseHeaders (which isn't in the interface).  I chose raw pointer and making shared at addReadFilter but lmk if you prefer casts.

Comment (snowp) [reply source/extensions/filters/network/http_connection_manager/config.cc:227]: Can't you just make a `shared_ptr<Http::ConnectionManagerImpl>` and then implicit casting will handle the upcast?

File: source/common/http/conn_manager_utility.cc (hunk 4 above)
Comment (snowp) [reply source/common/http/conn_manager_utility.cc]: wfm!

File: source/extensions/filters/network/http_connection_manager/config.cc (hunk 6 above)
Comment (alyssawilk) [reply source/extensions/filters/network/http_connection_manager/config.cc:227]: good call, on it!

**Review by snowp — APPROVED** [review]
Thanks!

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by alyssawilk** [comment]
ugh, so this doesn't work as advertised, because Envoy mobile sets up its config via ApiListener, and ApiListener is hard-coded to only read v3::HttpConnectionManager

I assume I'll need to do a follow-up to allow it to also parse EnvoyMobileHttpConnectionManager, but I'm also 1600+ diff patches into the envoy-mobile change to move from HttpConnectionManager to EnvoyMobileHttpConnectionManager and if anyone is willing to revisit a knob in the base HCM I'd love to not add more scut work here....

**Comment by mattklein123** [comment]
> I assume I'll need to do a follow-up to allow it to also parse EnvoyMobileHttpConnectionManager, but I'm also 1600+ diff patches into the envoy-mobile change to move from HttpConnectionManager to EnvoyMobileHttpConnectionManager and if anyone is willing to revisit a knob in the base HCM I'd love to not add more scut work here....

I'm really surprised it's this much of a diff? In general I think it's worth doing this work as I think we are going to need more options passed through that we don't necessarily want to be Envoy proxy side visible.

**Comment by alyssawilk** [comment]
It's mostly whitespace: they have many hard-coded configs, and along with adding the new type becasue no-proto-subclass one has to add config: and add 2 spaces to all the HCM config.  I've already done it so no skin off my back if we continue but it's a lot of churn I didn't anticipate.

**Pull request merged**